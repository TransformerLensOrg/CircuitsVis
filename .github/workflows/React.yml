name: React

on:
  push:
    branches: 
      - master
  pull_request:
    branches: 
      - master

env:
  working-directory: 'pysvelte'

jobs:
  test:
    name: "Test, Type Check & Lint"
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Install dependencies
      run: yarn
      working-directory: ${{env.working-directory}}
    - name: Jest
      run: yarn test
      working-directory: ${{env.working-directory}}
    - name: Lint
      run: yarn lint
      working-directory: ${{env.working-directory}}
    - name: Check types
      run: yarn typeCheck
      working-directory: ${{env.working-directory}}
