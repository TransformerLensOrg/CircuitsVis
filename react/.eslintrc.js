module.exports = {
  env: {
    browser: true,
    es6: true,
    node: true,
    jest: true
  },
  extends: [
    "airbnb-base",
    "airbnb-typescript",
    "plugin:prettier/recommended",
    "plugin:react/recommended",
    "plugin:react-hooks/recommended",
    "plugin:testing-library/react",
    "plugin:jest-dom/recommended",
    "plugin:storybook/recommended"
  ],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: ["./tsconfig.json"],
    ecmaVersion: 2018,
    ecmaFeatures: {
      jsx: true
    },
    sourceType: "module"
  },
  plugins: [
    "@typescript-eslint",
    "jest-dom",
    "jsx-a11y",
    "react",
    "testing-library"
  ],
  rules: {
    "import/prefer-default-export": "off",
    "import/no-extraneous-dependencies": ["error", { "devDependencies": ["**/*.test.*", "**/*.stories.*"] }],
    "react/jsx-filename-extension": "off",
    "react/require-default-props": "off", // Not needed with TS
    "react/react-in-jsx-scope": "off" // Esbuild injects this for us
  },
  settings: {
    react: {
      version: "detect"
    }
  }
};
