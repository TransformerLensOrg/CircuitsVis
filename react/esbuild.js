/* eslint-disable import/no-extraneous-dependencies */
const esbuild = require("esbuild");
const { externalGlobalPlugin } = require("esbuild-plugin-external-global");

/**
 * CDN IIFE Build (with React)
 */
esbuild.build({
  entryPoints: ["src/index.ts"],
  outfile: "dist/cdn/iife.js",
  bundle: true,
  target: "es6",
  minify: true,
  legalComments: "none",
  sourcemap: true,
  globalName: "CircuitsVis" // Components available as e.g. `CircuitsVis.Hello`
});

/**
 * CDN ESM Build (with React)
 */
esbuild.build({
  entryPoints: ["src/index.ts"],
  outfile: "dist/cdn/esm.js",
  bundle: true,
  target: "es2020",
  format: "esm",
  minify: true,
  platform: "browser",
  legalComments: "none",
  sourcemap: true
});

/**
 * CDN IIFE Build (without React)
 *
 * This allows the user to import and run alongside their own browser import of
 * React (whichever version that may be).
 */
esbuild.build({
  entryPoints: ["src/index.ts"],
  outfile: "dist/cdn/without-react.iife.js",
  bundle: true,
  target: "es6",
  minify: true,
  legalComments: "none",
  sourcemap: true,
  globalName: "CircuitsVis", // Components available as e.g. `CircuitsVis.Hello`
  plugins: [
    // Exclude React/ReactDom from the browser
    externalGlobalPlugin({
      react: "window.React",
      "react-dom": "window.ReactDOM"
    })
  ]
});

/**
 * CommonJS Version
 */
esbuild.build({
  entryPoints: ["src/index.ts"],
  outdir: "dist/commonjs/",
  external: ["./node_modules/*"],
  sourcemap: true,
  bundle: true,
  target: ["node12"],
  platform: "node",
  format: "cjs"
});
