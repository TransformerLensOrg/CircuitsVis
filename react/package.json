{
  "name": "circuitsvis",
  "description": "Mechanistic Interpretability Visualizations",
  "version": "1.35.0",
  "main": "dist/commonjs/index.js",
  "module": "dist/module/index.js",
  "unpkg": "dist/cdn/iife.js",
  "jsdelivr": "dist/cdn/iife.js",
  "source": "src/index.ts",
  "types": "dist/module/index.d.ts",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.esm.mjs",
      "require": "./dist/index.cjs.js"
    }
  },
  "scripts": {
    "test": "jest",
    "lint": "eslint src",
    "buildBrowser": "node ./esbuild.js",
    "buildNode": "tsc --project tsconfig.build.json",
    "build": "yarn buildBrowser && yarn buildNode",
    "typeCheck": "tsc --noEmit",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "repository": "https://github.com/alan-cooney/CircuitsVis",
  "author": "Alan Cooney",
  "license": "MIT",
  "dependencies": {
    "@tensorflow/tfjs": "^4.1.0",
    "chart.js": "^4.0.1",
    "chartjs-chart-matrix": "^1.3.0",
    "colord": "^2.9.3",
    "react-chartjs-2": "^5.0.1",
    "react-grid-system": "^8.1.6",
    "react-popper-tooltip": "^4.4.2",
    "tinycolor2": "^1.4.2"
  },
  "devDependencies": {
    "@storybook/addon-actions": "^6.5.14",
    "@storybook/addon-essentials": "^6.5.14",
    "@storybook/addon-interactions": "^6.5.14",
    "@storybook/addon-links": "^6.5.14",
    "@storybook/builder-webpack5": "^6.5.14",
    "@storybook/manager-webpack5": "^6.5.14",
    "@storybook/preset-typescript": "^3.0.0",
    "@storybook/react": "^6.5.14",
    "@storybook/testing-library": "^0.0.13",
    "@tensorflow/tfjs-node": "^4.1.0",
    "@testing-library/jest-dom": "5.16.5",
    "@testing-library/react": "^13.4.0",
    "@testing-library/react-hooks": "^8.0.1",
    "@testing-library/user-event": "14.4.3",
    "@types/jest": "29.2.4",
    "@types/node": "^18.11.11",
    "@types/react": "^18.0.26",
    "@types/testing-library__jest-dom": "^5.14.5",
    "@types/tinycolor2": "^1.4.3",
    "@typescript-eslint/eslint-plugin": "5.45.1",
    "@typescript-eslint/parser": "^5.45.1",
    "chromatic": "^6.12.0",
    "esbuild": "^0.16.1",
    "esbuild-plugin-external-global": "^1.0.1",
    "eslint": "8.29.0",
    "eslint-config-airbnb-base": "15.0.0",
    "eslint-config-airbnb-typescript": "17.0.0",
    "eslint-config-prettier": "8.5.0",
    "eslint-import-resolver-typescript": "3.5.2",
    "eslint-plugin-import": "2.26.0",
    "eslint-plugin-jest": "27.1.6",
    "eslint-plugin-jest-dom": "4.0.3",
    "eslint-plugin-jsx-a11y": "6.6.1",
    "eslint-plugin-prettier": "4.2.1",
    "eslint-plugin-react": "7.31.11",
    "eslint-plugin-react-hooks": "4.6.0",
    "eslint-plugin-storybook": "^0.6.8",
    "eslint-plugin-testing-library": "5.9.1",
    "jest": "^29.3.1",
    "jest-canvas-mock": "^2.4.0",
    "jest-environment-jsdom": "29.3.1",
    "prettier": "^2.8.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "testing-library": "^0.0.2",
    "ts-jest": "^29.0.3",
    "typescript": "^4.9.3"
  },
  "peerDependencies": {
    "react": "^16.8.0 || ^17 || ^18"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "transform": {
      "^.+\\.tsx?$": [
        "ts-jest"
      ]
    },
    "setupFiles": [
      "jest-canvas-mock"
    ]
  }
}
