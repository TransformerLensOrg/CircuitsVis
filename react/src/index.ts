export { AttentionPatterns } from "./attention/AttentionPatterns";
export { Hello } from "./examples/Hello";
export { render } from "./render-helper";
