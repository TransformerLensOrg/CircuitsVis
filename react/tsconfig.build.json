{
  "extends": "./tsconfig.json",
  "exclude": [
    "src/**/*.spec.ts",
    "src/**/*.test.ts",
    "src/**/*.test.tsx",
    "src/**/*.stories.ts",
    "src/**/*.stories.tsx",
    "esbuild.js"
  ]
}
