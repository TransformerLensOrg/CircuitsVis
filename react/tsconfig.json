{
  "compilerOptions": {
    "sourceMap": true,
    "module": "es2015",
    "target": "es2018",
    "moduleResolution": "node",
    "outDir": "./dist/module",
    "jsx": "react",
    "skipLibCheck": true,
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
  },
  "include": [
    "src",
    "esbuild.js"
  ],
  "exclude": [
    "node_modules",
  ],
}
