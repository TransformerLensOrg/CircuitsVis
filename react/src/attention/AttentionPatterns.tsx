import React, { useMemo, useState } from "react";
import { einsum, Rank, tensor, Tensor3D, Tensor4D } from "@tensorflow/tfjs";
import tinycolor from "tinycolor2";
import { AttentionImage } from "./components/AttentionImage";
import { Tokens, TokensView } from "./components/AttentionTokens";
import { useHoverLock } from "./components/useHoverLock";

/**
 * Color the attention values by heads
 *
 * We want attention values (re-ordered to [heads x dest_tokens x src_tokens]),
 * to be colored by each head (i.e. becoming [heads x dest_tokens x src_tokens x
 * rgb_color_channel]). This way, when outputting an image of just one attention
 * head it will be colored (by the specific hue assigned to that attention head)
 * rather than grayscale.
 *
 * Importantly, when outputting an image that averages
 * several attention heads we can then also average over the colors (so that we
 * can see for each destination-source token pair which head is most important).
 * For example, if the specific pair is very red, it suggests that the red
 * attention head is most important for this destination-source token combination.
 *
 * @param attentionInput Attention input as [dest_tokens x source_tokens x
 * heads] array (this is the format provided by the Python interface).
 *
 * @returns Tensor of the shape [heads x dest_tokens x src_tokens x
 * rgb_color_channel]
 */
export function colorAttentionTensors(attentionInput: number[][][]): Tensor4D {
  // Create a TensorFlow tensor from the attention data
  const attentionTensor = tensor<Rank.R3>(attentionInput); // [heads x dest_tokens x source_tokens]

  const attention = attentionTensor.arraySync() as number[][][];

  // Set the colors
  const colored = attention.map((head, headNumber) =>
    head.map((destination) =>
      destination.map((sourceAttention) => {
        // Color
        const attentionColor = tinycolor({
          h: (headNumber / attention.length) * 360, // Hue (degrees 0-360)
          s: 0.8, // Saturation (slightly off 100% to make less glaring)
          l: 1 - sourceAttention // Luminance (shows amount of attention)
        });

        // Return as a nested list in the format [red, green, blue]
        const { r, g, b } = attentionColor.toRgb();
        return [r, g, b];
      })
    )
  );

  return tensor(colored);
}

/**
 * Attention Patterns
 */
export function AttentionPatterns({
  tokens,
  attention
}: {
  /**
   * Array of tokens
   *
   * @example
   * ```typescript
   * ["Hello", "my", "name", "is"...]
   * ```
   */
  tokens: string[];
  /** Attention input as [dest_tokens x source_tokens x heads] (JSON stringified) */
  attention: number[][][];
}) {
  // State for the token view type
  const [tokensView, setTokensView] = useState<TokensView>(
    TokensView.DESTINATION_TO_SOURCE
  );

  // Attention head focussed state
  const {
    focused: focusedHead,
    onClick: onClickHead,
    onMouseEnter: onMouseEnterHead,
    onMouseLeave: onMouseLeaveHead
  } = useHoverLock();

  // State for which token is focussed
  const {
    focused: focussedToken,
    onClick: onClickToken,
    onMouseEnter: onMouseEnterToken,
    onMouseLeave: onMouseLeaveToken
  } = useHoverLock();

  // Color the attention values (by head)
  const coloredAttention = useMemo(
    () => colorAttentionTensors(attention),
    [attention]
  );
  const heads = coloredAttention.unstack<Tensor3D>(0);

  // Max attention color across all heads
  // This is helpful as we can see if, for example, only one or two colored
  // heads are focussing on a specific source token from a destination token.
  // To do this we re-arrange to put heads at the last dimension and then max
  // this (by color darkness, so min in terms of rgb values)
  const maxAttentionAcrossHeads = einsum("hdsc -> dsch", coloredAttention).min(
    3
  );

  // Get the focused head based on the state (selected/hovered)
  const focusedAttention =
    focusedHead === null ? maxAttentionAcrossHeads : heads[focusedHead];

  return (
    <div>
      <div style={{ display: "flex" }}>
        <div>
          <h4>Attention Patterns</h4>
          <AttentionImage coloredAttention={focusedAttention as any} />
        </div>

        <div style={{ marginLeft: 15 }}>
          <h4>
            Head selector
            <span style={{ fontWeight: "normal" }}>
              {" "}
              (hover to focus, click to lock)
            </span>
          </h4>
          <div style={{ display: "flex", flexWrap: "wrap" }}>
            {heads.map((head, headNumber) => (
              <figure
                key={headNumber}
                style={{
                  margin: 0,
                  marginRight: 15
                }}
                onClick={() => onClickHead(headNumber)}
                onMouseEnter={() => onMouseEnterHead(headNumber)}
                onMouseLeave={onMouseLeaveHead}
              >
                <AttentionImage
                  coloredAttention={head}
                  style={{ width: 80 }}
                  isSelected={headNumber === focusedHead}
                />
                <figcaption>Head {headNumber}</figcaption>
              </figure>
            ))}
          </div>
        </div>
      </div>

      <div className="tokens">
        <h4 style={{ display: "inline-block", marginRight: 15 }}>
          Tokens
          <span style={{ fontWeight: "normal" }}> (click to focus)</span>
        </h4>
        <select
          value={tokensView}
          onChange={(e) => setTokensView(e.target.value as any)}
        >
          <option value={TokensView.DESTINATION_TO_SOURCE}>
            Source ← Destination
          </option>
          <option value={TokensView.SOURCE_TO_DESTINATION}>
            Destination ← Source
          </option>
        </select>
        <div>
          <Tokens
            coloredAttention={coloredAttention}
            focusedHead={focusedHead}
            focusedToken={focussedToken}
            onClickToken={onClickToken}
            onMouseEnterToken={onMouseEnterToken}
            onMouseLeaveToken={onMouseLeaveToken}
            tokens={tokens}
            tokensView={tokensView}
          />
        </div>
      </div>
    </div>
  );
}
