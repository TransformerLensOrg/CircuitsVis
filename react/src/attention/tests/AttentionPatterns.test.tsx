import { setBackend, ones, Rank } from "@tensorflow/tfjs";
import { render, screen } from "@testing-library/react";
import React from "react";
import { AttentionPatterns, colorAttentionTensors } from "../AttentionPatterns";

beforeAll(() => {
  // Use the node backend whilst testing
  setBackend("cpu");
});

// Input with 2 tokens, 2 heads
const mockAttention = [
  [
    [1, 0],
    [0.5, 0.5]
  ],
  [
    [1, 0],
    [0.5, 0.5]
  ]
];

describe("colorAttentionTensors", () => {
  it("creates a tensor of the correct shape", () => {
    const res = colorAttentionTensors(mockAttention);
    expect(res.shape).toEqual([2, 2, 2, 3]);
  });

  it("creates colors for heads matching the snapshot", () => {
    const res = colorAttentionTensors(mockAttention);
    expect(res.arraySync()).toMatchSnapshot();
  });
});

describe("AttentionPatterns", () => {
  it("renders", () => {
    const tokens = ["A", "B", "C", "D"];
    const attention = ones<Rank.R3>([4, 4, 16]);

    render(
      <AttentionPatterns tokens={tokens} attention={attention.arraySync()} />
    );

    // Check the header text loads
    screen.getByText("Attention Patterns");
  });
});
