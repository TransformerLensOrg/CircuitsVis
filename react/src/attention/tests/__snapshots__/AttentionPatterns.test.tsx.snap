// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorAttentionTensors creates colors for heads matching the snapshot 1`] = `
[
  [
    [
      [
        0,
        0,
        0,
      ],
      [
        230,
        25,
        25,
      ],
    ],
    [
      [
        0,
        0,
        0,
      ],
      [
        230,
        25,
        25,
      ],
    ],
  ],
  [
    [
      [
        255,
        255,
        255,
      ],
      [
        25,
        229,
        230,
      ],
    ],
    [
      [
        255,
        255,
        255,
      ],
      [
        25,
        229,
        230,
      ],
    ],
  ],
]
`;
